
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Principal component analysis &#8212; Population genomics in practice</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=62ba249389abaaa9ffc34bf36a076bdc1d65ee18" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=f31d14ad54b65d19161ba51d4ffff3a77ae00456"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="PCA with sgkit" href="pca/sgkit.html" />
    <link rel="prev" title="Describing variation" href="describing_variation.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
      
      
      <h1 class="site-logo" id="site-title">Population genomics in practice</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Population Genomics in Practice exercises
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="schedule.html">
   Schedule
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Day 0: Preparation
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="preparation.html">
   Preparation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Day 1: Models of evolution
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="wrightfisher.html">
   Wright-Fisher model
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Day 2: Variant calling
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="variantcalling.html">
   Variant calling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="variantfiltering.html">
   Variantfiltering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="genotype_likelihoods.html">
   Genotype likelihoods and low-coverage data
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Day 3: Describing variation and population structure
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="describing_variation.html">
   Describing variation
  </a>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="current reference internal" href="#">
   Principal component analysis
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="pca/sgkit.html">
     PCA with sgkit
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="pca/plink.html">
     PCA with plink
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="admixture.html">
   Admixture
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Day 4: Demographic inference
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="demographic_inference.html">
   Demographic inference
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="fstatistics.html">
   f-statistics
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Day 5: Selection
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="selection.html">
   Selection
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  About
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="data.html">
   Data
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Bibliography
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="references.html">
   References
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/NBISweden/pgip-exercises"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/NBISweden/pgip-exercises/issues/new?title=Issue%20on%20page%20%2Fld_pruning_pca.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="_sources/ld_pruning_pca.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download notebook file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-code"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        <a href="_sources/ld_pruning_pca.md"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.md</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-preparation">
   Data preparation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#linkage-disequilibrium-structure">
   Linkage disequilibrium structure
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#linkage-disequilibrium-pruning">
   Linkage disequilibrium pruning
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pca">
   PCA
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Principal component analysis</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-preparation">
   Data preparation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#linkage-disequilibrium-structure">
   Linkage disequilibrium structure
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#linkage-disequilibrium-pruning">
   Linkage disequilibrium pruning
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pca">
   PCA
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <div class="cell tag_hide-output docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">bokeh.plotting</span> <span class="kn">import</span> <span class="n">figure</span><span class="p">,</span> <span class="n">show</span><span class="p">,</span> <span class="n">output_notebook</span>
<span class="kn">from</span> <span class="nn">pgip</span> <span class="kn">import</span> <span class="n">utils</span>
<span class="n">output_notebook</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="bk-root">
        <a href="https://bokeh.org" target="_blank" class="bk-logo bk-logo-small bk-logo-notebook"></a>
        <span id="1002">Loading BokehJS ...</span>
    </div>
</div><script type="application/javascript">(function(root) {
  function now() {
    return new Date();
  }

  const force = true;

  if (typeof root._bokeh_onload_callbacks === "undefined" || force === true) {
    root._bokeh_onload_callbacks = [];
    root._bokeh_is_loading = undefined;
  }

const JS_MIME_TYPE = 'application/javascript';
  const HTML_MIME_TYPE = 'text/html';
  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';
  const CLASS_NAME = 'output_bokeh rendered_html';

  /**
   * Render data to the DOM node
   */
  function render(props, node) {
    const script = document.createElement("script");
    node.appendChild(script);
  }

  /**
   * Handle when an output is cleared or removed
   */
  function handleClearOutput(event, handle) {
    const cell = handle.cell;

    const id = cell.output_area._bokeh_element_id;
    const server_id = cell.output_area._bokeh_server_id;
    // Clean up Bokeh references
    if (id != null && id in Bokeh.index) {
      Bokeh.index[id].model.document.clear();
      delete Bokeh.index[id];
    }

    if (server_id !== undefined) {
      // Clean up Bokeh references
      const cmd_clean = "from bokeh.io.state import curstate; print(curstate().uuid_to_server['" + server_id + "'].get_sessions()[0].document.roots[0]._id)";
      cell.notebook.kernel.execute(cmd_clean, {
        iopub: {
          output: function(msg) {
            const id = msg.content.text.trim();
            if (id in Bokeh.index) {
              Bokeh.index[id].model.document.clear();
              delete Bokeh.index[id];
            }
          }
        }
      });
      // Destroy server and session
      const cmd_destroy = "import bokeh.io.notebook as ion; ion.destroy_server('" + server_id + "')";
      cell.notebook.kernel.execute(cmd_destroy);
    }
  }

  /**
   * Handle when a new output is added
   */
  function handleAddOutput(event, handle) {
    const output_area = handle.output_area;
    const output = handle.output;

    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only
    if ((output.output_type != "display_data") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {
      return
    }

    const toinsert = output_area.element.find("." + CLASS_NAME.split(' ')[0]);

    if (output.metadata[EXEC_MIME_TYPE]["id"] !== undefined) {
      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];
      // store reference to embed id on output_area
      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE]["id"];
    }
    if (output.metadata[EXEC_MIME_TYPE]["server_id"] !== undefined) {
      const bk_div = document.createElement("div");
      bk_div.innerHTML = output.data[HTML_MIME_TYPE];
      const script_attrs = bk_div.children[0].attributes;
      for (let i = 0; i < script_attrs.length; i++) {
        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);
        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent
      }
      // store reference to server id on output_area
      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE]["server_id"];
    }
  }

  function register_renderer(events, OutputArea) {

    function append_mime(data, metadata, element) {
      // create a DOM node to render to
      const toinsert = this.create_output_subarea(
        metadata,
        CLASS_NAME,
        EXEC_MIME_TYPE
      );
      this.keyboard_manager.register_events(toinsert);
      // Render to node
      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};
      render(props, toinsert[toinsert.length - 1]);
      element.append(toinsert);
      return toinsert
    }

    /* Handle when an output is cleared or removed */
    events.on('clear_output.CodeCell', handleClearOutput);
    events.on('delete.Cell', handleClearOutput);

    /* Handle when a new output is added */
    events.on('output_added.OutputArea', handleAddOutput);

    /**
     * Register the mime type and append_mime function with output_area
     */
    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {
      /* Is output safe? */
      safe: true,
      /* Index of renderer in `output_area.display_order` */
      index: 0
    });
  }

  // register the mime type if in Jupyter Notebook environment and previously unregistered
  if (root.Jupyter !== undefined) {
    const events = require('base/js/events');
    const OutputArea = require('notebook/js/outputarea').OutputArea;

    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {
      register_renderer(events, OutputArea);
    }
  }
  if (typeof (root._bokeh_timeout) === "undefined" || force === true) {
    root._bokeh_timeout = Date.now() + 5000;
    root._bokeh_failed_load = false;
  }

  const NB_LOAD_WARNING = {'data': {'text/html':
     "<div style='background-color: #fdd'>\n"+
     "<p>\n"+
     "BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \n"+
     "may be due to a slow or bad network connection. Possible fixes:\n"+
     "</p>\n"+
     "<ul>\n"+
     "<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\n"+
     "<li>use INLINE resources instead, as so:</li>\n"+
     "</ul>\n"+
     "<code>\n"+
     "from bokeh.resources import INLINE\n"+
     "output_notebook(resources=INLINE)\n"+
     "</code>\n"+
     "</div>"}};

  function display_loaded() {
    const el = document.getElementById("1002");
    if (el != null) {
      el.textContent = "BokehJS is loading...";
    }
    if (root.Bokeh !== undefined) {
      if (el != null) {
        el.textContent = "BokehJS " + root.Bokeh.version + " successfully loaded.";
      }
    } else if (Date.now() < root._bokeh_timeout) {
      setTimeout(display_loaded, 100)
    }
  }

  function run_callbacks() {
    try {
      root._bokeh_onload_callbacks.forEach(function(callback) {
        if (callback != null)
          callback();
      });
    } finally {
      delete root._bokeh_onload_callbacks
    }
    console.debug("Bokeh: all callbacks have finished");
  }

  function load_libs(css_urls, js_urls, callback) {
    if (css_urls == null) css_urls = [];
    if (js_urls == null) js_urls = [];

    root._bokeh_onload_callbacks.push(callback);
    if (root._bokeh_is_loading > 0) {
      console.debug("Bokeh: BokehJS is being loaded, scheduling callback at", now());
      return null;
    }
    if (js_urls == null || js_urls.length === 0) {
      run_callbacks();
      return null;
    }
    console.debug("Bokeh: BokehJS not loaded, scheduling load and callback at", now());
    root._bokeh_is_loading = css_urls.length + js_urls.length;

    function on_load() {
      root._bokeh_is_loading--;
      if (root._bokeh_is_loading === 0) {
        console.debug("Bokeh: all BokehJS libraries/stylesheets loaded");
        run_callbacks()
      }
    }

    function on_error(url) {
      console.error("failed to load " + url);
    }

    for (let i = 0; i < css_urls.length; i++) {
      const url = css_urls[i];
      const element = document.createElement("link");
      element.onload = on_load;
      element.onerror = on_error.bind(null, url);
      element.rel = "stylesheet";
      element.type = "text/css";
      element.href = url;
      console.debug("Bokeh: injecting link tag for BokehJS stylesheet: ", url);
      document.body.appendChild(element);
    }

    for (let i = 0; i < js_urls.length; i++) {
      const url = js_urls[i];
      const element = document.createElement('script');
      element.onload = on_load;
      element.onerror = on_error.bind(null, url);
      element.async = false;
      element.src = url;
      console.debug("Bokeh: injecting script tag for BokehJS library: ", url);
      document.head.appendChild(element);
    }
  };

  function inject_raw_css(css) {
    const element = document.createElement("style");
    element.appendChild(document.createTextNode(css));
    document.body.appendChild(element);
  }

  const js_urls = ["https://cdn.bokeh.org/bokeh/release/bokeh-2.4.3.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.3.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.3.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.3.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.3.min.js"];
  const css_urls = [];

  const inline_js = [    function(Bokeh) {
      Bokeh.set_log_level("info");
    },
function(Bokeh) {
    }
  ];

  function run_inline_js() {
    if (root.Bokeh !== undefined || force === true) {
          for (let i = 0; i < inline_js.length; i++) {
      inline_js[i].call(root, root.Bokeh);
    }
if (force === true) {
        display_loaded();
      }} else if (Date.now() < root._bokeh_timeout) {
      setTimeout(run_inline_js, 100);
    } else if (!root._bokeh_failed_load) {
      console.log("Bokeh: BokehJS failed to load within specified timeout.");
      root._bokeh_failed_load = true;
    } else if (force !== true) {
      const cell = $(document.getElementById("1002")).parents('.cell').data().cell;
      cell.output_area.append_execute_result(NB_LOAD_WARNING)
    }
  }

  if (root._bokeh_is_loading === 0) {
    console.debug("Bokeh: BokehJS loaded, going straight to plotting");
    run_inline_js();
  } else {
    load_libs(css_urls, js_urls, function() {
      console.debug("Bokeh: BokehJS plotting callback run at", now());
      run_inline_js();
    });
  }
}(window));</script></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>We assume that the input data has already been filtered and pre-processed to obtain a high-quality variant call set. See <a class="reference internal" href="variantfiltering.html#sec-variantfiltering"><span class="std std-ref">Variantfiltering</span></a>.</p></li>
<li><p>sgkit pca method does not support missing data - add example?</p></li>
</ul>
</div>
<section class="tex2jax_ignore mathjax_ignore" id="principal-component-analysis">
<span id="sec-ld-pruning-pca"></span><h1>Principal component analysis<a class="headerlink" href="#principal-component-analysis" title="Permalink to this headline">#</a></h1>
<p>Variant data is high-dimensional and some sort of dimensionality
reduction is often employed to provide a condensed overview of the
data and as a means to identify potential population structure.
Principal component analysis (PCA) is one common technique usually
applied to genetic variation data (e.g.
<span id="id1">[<a class="reference internal" href="references.html#id8" title="Nick Patterson, Alkes L. Price, and David Reich. Population Structure and Eigenanalysis. PLOS Genetics, 2(12):e190, December 2006. doi:10.1371/journal.pgen.0020190.">Patterson <em>et al.</em>, 2006</a>]</span>).</p>
<p>For large data sets, it is common practice to first prune the data to
remove correlation between data points. In this exercise, you will
prune the data using either <a class="reference internal" href="pca/plink.html#sec-pca-with-plink"><span class="std std-ref">plink</span></a> or
<a class="reference internal" href="pca/sgkit.html#sec-pca-with-sgkit"><span class="std std-ref">sgkit</span></a> code snippets.</p>
<section id="data-preparation">
<h2>Data preparation<a class="headerlink" href="#data-preparation" title="Permalink to this headline">#</a></h2>
<div class="warning admonition">
<p class="admonition-title">FIXME</p>
<p>sgkit is likely too cumbersome? Keep as reference for now, use EIGENSOFT or plink for pruning and pca.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sgkit</span> <span class="k">as</span> <span class="nn">sg</span>
<span class="kn">from</span> <span class="nn">sgkit.io</span> <span class="kn">import</span> <span class="n">vcf</span> <span class="k">as</span> <span class="n">sgvcf</span>
</pre></div>
</div>
</div>
</div>
<p>Then make output directories and convert vcf to the zarr format.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">outdir</span> <span class="o">=</span> <span class="s2">&quot;exercises/ld_pruning&quot;</span>
<span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">vcf</span> <span class="o">=</span> <span class="s2">&quot;data/ooa/ooa.gatk.vcf.gz&quot;</span>
<span class="n">zarr</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="s2">&quot;ooa.gatk.zarr&quot;</span><span class="p">)</span>
<span class="n">sgvcf</span><span class="o">.</span><span class="n">vcf_to_zarr</span><span class="p">(</span><span class="n">vcf</span><span class="p">,</span> <span class="n">zarr</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We load the data and remember the original chunk size (needed
internally for sgkit):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">myst_nb</span> <span class="kn">import</span> <span class="n">glue</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="n">zarr</span><span class="p">)</span>
<span class="n">original_chunk_size</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">chunks</span><span class="p">[</span><span class="s2">&quot;variants&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">glue</span><span class="p">(</span><span class="s2">&quot;nvar&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">variants</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>648
</pre></div>
</div>
</div>
</div>
</section>
<section id="linkage-disequilibrium-structure">
<h2>Linkage disequilibrium structure<a class="headerlink" href="#linkage-disequilibrium-structure" title="Permalink to this headline">#</a></h2>
<p>Before pruning the data, we make a plot of the LD structure. We will
use <code class="docutils literal notranslate"><span class="pre">python</span></code> for plotting and first load the necessary libraries:</p>
</section>
<section id="linkage-disequilibrium-pruning">
<h2>Linkage disequilibrium pruning<a class="headerlink" href="#linkage-disequilibrium-pruning" title="Permalink to this headline">#</a></h2>
<p>Depending on the number of variants, it may be necessary to work on a
subset of variants. Here with only <span class="pasted-inline"><code class="output text_plain docutils literal notranslate"><span class="pre">648</span></code></span> variants we skip this
step.</p>
<p>For pruning, we first need to calculate the dosage:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ds</span><span class="p">[</span><span class="s1">&#39;dosage&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ds</span><span class="p">[</span><span class="s1">&#39;call_genotype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="s1">&#39;ploidy&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>and then divide into windows by variants:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># ds = sg.window_by_variant(ds, size=200)</span>
<span class="c1"># print(ds)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="pca">
<h2>PCA<a class="headerlink" href="#pca" title="Permalink to this headline">#</a></h2>
<p>Before proceeding with the pca, we need to compute quality control
variant statistics from genotype calls followed by calculation of and
filtering on allele frequencies.</p>
<!-- See https://github.com/pystatgen/sgkit/issues/752 -->
<!-- All the pca code is based on that example --><div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">variant_stats</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>

<span class="n">ds</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="o">**</span><span class="n">ds</span><span class="p">[[</span><span class="s2">&quot;variant_allele_frequency&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">compute</span><span class="p">())</span><span class="o">.</span><span class="n">pipe</span><span class="p">(</span>
<span class="k">lambda</span> <span class="n">ds</span><span class="p">:</span> <span class="n">ds</span><span class="o">.</span><span class="n">sel</span><span class="p">(</span><span class="n">variants</span><span class="o">=</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">variant_allele_frequency</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.01</span><span class="p">))</span>
<span class="p">)</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">chunk</span><span class="p">(</span><span class="n">original_chunk_size</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now we can perform pca.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Transform genotype call into number of non-reference alleles</span>
<span class="n">ds_pca</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">count_call_alternate_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>
<span class="c1"># Ensure data has positive variance across samples</span>
<span class="n">variant_mask</span> <span class="o">=</span> <span class="p">((</span><span class="n">ds_pca</span><span class="o">.</span><span class="n">call_alternate_allele_count</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="s2">&quot;samples&quot;</span><span class="p">))</span> <span class="o">|</span> <span class="p">(</span>
<span class="n">ds_pca</span><span class="o">.</span><span class="n">call_alternate_allele_count</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="s2">&quot;samples&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mf">0.0</span>
<span class="p">)</span>

<span class="n">ds_pca</span> <span class="o">=</span> <span class="n">ds_pca</span><span class="o">.</span><span class="n">sel</span><span class="p">(</span><span class="n">variants</span><span class="o">=~</span><span class="n">variant_mask</span><span class="p">)</span>
<span class="n">ds_pca</span><span class="p">[</span><span class="s2">&quot;call_alternate_allele_count&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ds_pca</span><span class="o">.</span><span class="n">call_alternate_allele_count</span><span class="o">.</span><span class="n">chunk</span><span class="p">(</span>
<span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">ds_pca</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">ds_pca</span><span class="p">,</span> <span class="n">n_components</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We convert the data structure to a pandas dataframe and add various
metadata, such as sample and population names.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">explained</span> <span class="o">=</span> <span class="n">ds_pca</span><span class="o">.</span><span class="n">sample_pca_explained_variance_ratio</span><span class="o">.</span><span class="n">values</span> <span class="o">*</span> <span class="mi">100</span>
<span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
<span class="n">ds_pca</span><span class="o">.</span><span class="n">sample_pca_projection</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
<span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="o">.</span><span class="n">pivot</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="s2">&quot;samples&quot;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s2">&quot;components&quot;</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span>
<span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
<span class="n">df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">df</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ds_pca</span><span class="o">.</span><span class="n">sample_id</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">df</span><span class="p">[</span><span class="s2">&quot;population&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">colors</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;population&quot;</span><span class="p">]),</span> <span class="n">utils</span><span class="o">.</span><span class="n">_get_palette</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">3</span><span class="p">)))</span>
<span class="n">df</span><span class="p">[</span><span class="s2">&quot;color&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">colors</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;population&quot;</span><span class="p">]]</span>
</pre></div>
</div>
</div>
</div>
<p>For the purpose of this exercise, we want to also investigate the
effect of sequencing coverage.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">io</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="n">output</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">check_output</span><span class="p">(</span><span class="s2">&quot;samtools depth -a data/ooa/*.bam&quot;</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">coverage</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">(</span><span class="n">output</span><span class="o">.</span><span class="n">decode</span><span class="p">()),</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">df</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">coverage</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/tmp/ipykernel_6026/465319288.py:5: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with &#39;numeric_only=None&#39;) is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  coverage = pd.read_table(io.StringIO(output.decode()), header=None).mean(axis=0)
</pre></div>
</div>
</div>
</div>
<p>Finally, we can plot the pca with bokeh.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pgip.plotting</span> <span class="kn">import</span> <span class="n">bokeh_plot_pca</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">bokeh_plot_pca</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">explained</span><span class="p">,</span> <span class="n">ncomp</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mi">400</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">400</span><span class="p">)</span>
<span class="n">show</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div class="bk-root" id="b1fbb2f8-0bdc-4c89-b2f2-c618e9aa6f56" data-root-id="1207"></div>
</div><script type="application/javascript">(function(root) {
  function embed_document(root) {
  const docs_json = {"f4e390b0-9777-4446-8ec9-182b8a0276c5":{"defs":[],"roots":{"references":[{"attributes":{"children":[{"id":"1206"},{"id":"1204"}]},"id":"1207","type":"Column"},{"attributes":{"bottom_units":"screen","coordinates":null,"fill_alpha":0.5,"fill_color":"lightgrey","group":null,"left_units":"screen","level":"overlay","line_alpha":1.0,"line_color":"black","line_dash":[4,4],"line_width":2,"right_units":"screen","syncable":false,"top_units":"screen"},"id":"1141","type":"BoxAnnotation"},{"attributes":{},"id":"1164","type":"UnionRenderers"},{"attributes":{"fill_alpha":{"value":0.1},"fill_color":{"field":"color"},"hatch_alpha":{"value":0.1},"hatch_color":{"field":"color"},"line_alpha":{"value":0.1},"size":{"value":15},"x":{"field":"PC2"},"y":{"field":"PC3"}},"id":"1153","type":"Circle"},{"attributes":{"index":4,"label":{"value":"CHB"},"renderers":[{"id":"1155"}]},"id":"1169","type":"LegendItem"},{"attributes":{},"id":"1165","type":"Selection"},{"attributes":{"children":[[{"id":"1004"},0,0],[{"id":"1060"},0,1],[{"id":"1116"},1,0]]},"id":"1204","type":"GridBox"},{"attributes":{"toolbars":[{"id":"1031"},{"id":"1087"},{"id":"1143"}],"tools":[{"id":"1023"},{"id":"1024"},{"id":"1025"},{"id":"1026"},{"id":"1027"},{"id":"1028"},{"id":"1030"},{"id":"1079"},{"id":"1080"},{"id":"1081"},{"id":"1082"},{"id":"1083"},{"id":"1084"},{"id":"1086"},{"id":"1135"},{"id":"1136"},{"id":"1137"},{"id":"1138"},{"id":"1139"},{"id":"1140"},{"id":"1142"}]},"id":"1205","type":"ProxyToolbar"},{"attributes":{"tools":[{"id":"1135"},{"id":"1136"},{"id":"1137"},{"id":"1138"},{"id":"1139"},{"id":"1140"},{"id":"1142"}]},"id":"1143","type":"Toolbar"},{"attributes":{"source":{"id":"1115"}},"id":"1156","type":"CDSView"},{"attributes":{"coordinates":null,"data_source":{"id":"1115"},"glyph":{"id":"1152"},"group":null,"hover_glyph":null,"muted_glyph":{"id":"1154"},"nonselection_glyph":{"id":"1153"},"view":{"id":"1156"}},"id":"1155","type":"GlyphRenderer"},{"attributes":{"fill_alpha":{"value":0.2},"fill_color":{"field":"color"},"hatch_alpha":{"value":0.2},"hatch_color":{"field":"color"},"line_alpha":{"value":0.2},"size":{"value":15},"x":{"field":"PC2"},"y":{"field":"PC3"}},"id":"1154","type":"Circle"},{"attributes":{"coordinates":null,"group":null,"items":[{"id":"1168"},{"id":"1169"},{"id":"1170"}]},"id":"1167","type":"Legend"},{"attributes":{"toolbar":{"id":"1205"},"toolbar_location":"above"},"id":"1206","type":"ToolbarBox"},{"attributes":{"below":[{"id":"1015"}],"center":[{"id":"1018"},{"id":"1022"},{"id":"1055"}],"height":400,"left":[{"id":"1019"}],"renderers":[{"id":"1043"}],"right":[{"id":"1055"}],"title":{"id":"1005"},"toolbar":{"id":"1031"},"toolbar_location":null,"width":400,"x_range":{"id":"1007"},"x_scale":{"id":"1011"},"y_range":{"id":"1009"},"y_scale":{"id":"1013"}},"id":"1004","subtype":"Figure","type":"Plot"},{"attributes":{"index":0,"label":{"value":"CEU"},"renderers":[{"id":"1043"}]},"id":"1056","type":"LegendItem"},{"attributes":{"below":[{"id":"1071"}],"center":[{"id":"1074"},{"id":"1078"},{"id":"1111"}],"height":400,"left":[{"id":"1075"}],"renderers":[{"id":"1099"}],"right":[{"id":"1111"}],"title":{"id":"1061"},"toolbar":{"id":"1087"},"toolbar_location":null,"width":400,"x_range":{"id":"1063"},"x_scale":{"id":"1067"},"y_range":{"id":"1065"},"y_scale":{"id":"1069"}},"id":"1060","subtype":"Figure","type":"Plot"},{"attributes":{},"id":"1160","type":"AllLabels"},{"attributes":{},"id":"1083","type":"ResetTool"},{"attributes":{},"id":"1011","type":"LinearScale"},{"attributes":{"index":4,"label":{"value":"CHB"},"renderers":[{"id":"1043"}]},"id":"1057","type":"LegendItem"},{"attributes":{"data":{"PC1":{"__ndarray__":"o9RFQdd2BkHQxztBDM2gwQPUWcGPDa3AMfIXwYz0Az/Yet9AAtcJQQ==","dtype":"float32","order":"little","shape":[10]},"PC10":{"__ndarray__":"AAD0tQAAB7YAABq2AAAWtgAAMLYAAMS1AAAUtgAA+LUAwAW2AIACtg==","dtype":"float32","order":"little","shape":[10]},"PC2":{"__ndarray__":"o4nNwJpKhsCgvqu+azLQwHP1BcE2vL5B+uo7QLY7dcB+rpjAehP1QA==","dtype":"float32","order":"little","shape":[10]},"PC3":{"__ndarray__":"0d3BwIKKT79jjDJAoLw1vhO43cCH9SrBlD+WQf6eqsD/jJG/9u4YQQ==","dtype":"float32","order":"little","shape":[10]},"PC4":{"__ndarray__":"XTNWwaPDi8BH7VC/oU4bwRgexEDJaR7ABF5Lvy7agUEc2KJABieDQA==","dtype":"float32","order":"little","shape":[10]},"PC5":{"__ndarray__":"O8siQHqPpUAwU/u//vYSwQD65LzP1UZAFYgzQaokBkBMquU/siJswQ==","dtype":"float32","order":"little","shape":[10]},"PC6":{"__ndarray__":"Kb3dQHLSIj9MUVHAEMedvylA8b6yo7G/neXnP6b1D0HcdHLBzAlJQA==","dtype":"float32","order":"little","shape":[10]},"PC7":{"__ndarray__":"HteRwLrMRkEtGXjAMDqQwC67CEGURI++csbPvyq90MDiBHPAqTOGQA==","dtype":"float32","order":"little","shape":[10]},"PC8":{"__ndarray__":"OW58wESzjr+kv3NBwzcgv7ZOekAgnIE8SOo/vzKDsL8MieDAJ5aKwA==","dtype":"float32","order":"little","shape":[10]},"PC9":{"__ndarray__":"XjjeQJ/yHcEVI0q/OmfOwC5rGUGoNT++oecVQHjgq8BRMF8/zgQ7QA==","dtype":"float32","order":"little","shape":[10]},"color":["#8dd3c7","#8dd3c7","#8dd3c7","#8dd3c7","#ffed6f","#ffed6f","#ffed6f","#d6ca66","#d6ca66","#d6ca66"],"coverage":{"__ndarray__":"AAAAAAAA+H8AAAAAAAD4f5nwS/286SBAK01KQbc3JEDogvqWOb0lQFMFo5I6kT9AcEIhAg5BJkAzbf/KShMeQFXZd0XwHyVAkwA1tWxNIkA=","dtype":"float64","order":"little","shape":[10]},"index":[0,1,2,3,4,5,6,7,8,9],"population":["CEU","CEU","CEU","CEU","CHB","CHB","CHB","YRI","YRI","YRI"],"sample":["CEU-0","CEU-1","CEU-2","CEU-3","CHB-0","CHB-1","CHB-2","YRI-0","YRI-1","YRI-2"]},"selected":{"id":"1053"},"selection_policy":{"id":"1052"}},"id":"1003","type":"ColumnDataSource"},{"attributes":{},"id":"1162","type":"BasicTickFormatter"},{"attributes":{"index":7,"label":{"value":"YRI"},"renderers":[{"id":"1043"}]},"id":"1058","type":"LegendItem"},{"attributes":{"bottom_units":"screen","coordinates":null,"fill_alpha":0.5,"fill_color":"lightgrey","group":null,"left_units":"screen","level":"overlay","line_alpha":1.0,"line_color":"black","line_dash":[4,4],"line_width":2,"right_units":"screen","syncable":false,"top_units":"screen"},"id":"1085","type":"BoxAnnotation"},{"attributes":{},"id":"1027","type":"ResetTool"},{"attributes":{},"id":"1163","type":"AllLabels"},{"attributes":{"axis_label":"PC1 (22.7%)","coordinates":null,"formatter":{"id":"1050"},"group":null,"major_label_policy":{"id":"1051"},"ticker":{"id":"1016"}},"id":"1015","type":"LinearAxis"},{"attributes":{},"id":"1063","type":"DataRange1d"},{"attributes":{"fill_alpha":{"value":0.8},"fill_color":{"field":"color"},"hatch_alpha":{"value":0.8},"hatch_color":{"field":"color"},"line_alpha":{"value":0.8},"size":{"value":15},"x":{"field":"PC1"},"y":{"field":"PC3"}},"id":"1096","type":"Circle"},{"attributes":{},"id":"1069","type":"LinearScale"},{"attributes":{"overlay":{"id":"1029"}},"id":"1025","type":"BoxZoomTool"},{"attributes":{},"id":"1026","type":"SaveTool"},{"attributes":{},"id":"1159","type":"BasicTickFormatter"},{"attributes":{},"id":"1024","type":"WheelZoomTool"},{"attributes":{"index":0,"label":{"value":"CEU"},"renderers":[{"id":"1155"}]},"id":"1168","type":"LegendItem"},{"attributes":{},"id":"1079","type":"PanTool"},{"attributes":{"axis":{"id":"1019"},"coordinates":null,"dimension":1,"group":null,"ticker":null},"id":"1022","type":"Grid"},{"attributes":{"fill_alpha":{"value":0.8},"fill_color":{"field":"color"},"hatch_alpha":{"value":0.8},"hatch_color":{"field":"color"},"line_alpha":{"value":0.8},"size":{"value":15},"x":{"field":"PC1"},"y":{"field":"PC2"}},"id":"1040","type":"Circle"},{"attributes":{},"id":"1020","type":"BasicTicker"},{"attributes":{"index":7,"label":{"value":"YRI"},"renderers":[{"id":"1155"}]},"id":"1170","type":"LegendItem"},{"attributes":{},"id":"1023","type":"PanTool"},{"attributes":{"overlay":{"id":"1085"}},"id":"1081","type":"BoxZoomTool"},{"attributes":{},"id":"1082","type":"SaveTool"},{"attributes":{},"id":"1084","type":"HelpTool"},{"attributes":{},"id":"1007","type":"DataRange1d"},{"attributes":{"axis":{"id":"1015"},"coordinates":null,"group":null,"ticker":null},"id":"1018","type":"Grid"},{"attributes":{"axis":{"id":"1075"},"coordinates":null,"dimension":1,"group":null,"ticker":null},"id":"1078","type":"Grid"},{"attributes":{},"id":"1076","type":"BasicTicker"},{"attributes":{"callback":null,"tooltips":[["sample","@sample"],["population","@population"],["coverage","@coverage"]]},"id":"1030","type":"HoverTool"},{"attributes":{"axis_label":"PC3 (13.0%)","coordinates":null,"formatter":{"id":"1103"},"group":null,"major_label_policy":{"id":"1104"},"ticker":{"id":"1076"}},"id":"1075","type":"LinearAxis"},{"attributes":{"coordinates":null,"group":null,"text":"PC1 vs PC2"},"id":"1005","type":"Title"},{"attributes":{"axis":{"id":"1071"},"coordinates":null,"group":null,"ticker":null},"id":"1074","type":"Grid"},{"attributes":{"callback":null,"tooltips":[["sample","@sample"],["population","@population"],["coverage","@coverage"]]},"id":"1086","type":"HoverTool"},{"attributes":{"axis_label":"PC2 (16.1%)","coordinates":null,"formatter":{"id":"1047"},"group":null,"major_label_policy":{"id":"1048"},"ticker":{"id":"1020"}},"id":"1019","type":"LinearAxis"},{"attributes":{},"id":"1080","type":"WheelZoomTool"},{"attributes":{},"id":"1013","type":"LinearScale"},{"attributes":{},"id":"1067","type":"LinearScale"},{"attributes":{},"id":"1016","type":"BasicTicker"},{"attributes":{"axis_label":"PC1 (22.7%)","coordinates":null,"formatter":{"id":"1106"},"group":null,"major_label_policy":{"id":"1107"},"ticker":{"id":"1072"}},"id":"1071","type":"LinearAxis"},{"attributes":{},"id":"1009","type":"DataRange1d"},{"attributes":{},"id":"1065","type":"DataRange1d"},{"attributes":{},"id":"1028","type":"HelpTool"},{"attributes":{},"id":"1072","type":"BasicTicker"},{"attributes":{"coordinates":null,"group":null,"text":"PC1 vs PC3"},"id":"1061","type":"Title"},{"attributes":{},"id":"1108","type":"UnionRenderers"},{"attributes":{"data":{"PC1":{"__ndarray__":"o9RFQdd2BkHQxztBDM2gwQPUWcGPDa3AMfIXwYz0Az/Yet9AAtcJQQ==","dtype":"float32","order":"little","shape":[10]},"PC10":{"__ndarray__":"AAD0tQAAB7YAABq2AAAWtgAAMLYAAMS1AAAUtgAA+LUAwAW2AIACtg==","dtype":"float32","order":"little","shape":[10]},"PC2":{"__ndarray__":"o4nNwJpKhsCgvqu+azLQwHP1BcE2vL5B+uo7QLY7dcB+rpjAehP1QA==","dtype":"float32","order":"little","shape":[10]},"PC3":{"__ndarray__":"0d3BwIKKT79jjDJAoLw1vhO43cCH9SrBlD+WQf6eqsD/jJG/9u4YQQ==","dtype":"float32","order":"little","shape":[10]},"PC4":{"__ndarray__":"XTNWwaPDi8BH7VC/oU4bwRgexEDJaR7ABF5Lvy7agUEc2KJABieDQA==","dtype":"float32","order":"little","shape":[10]},"PC5":{"__ndarray__":"O8siQHqPpUAwU/u//vYSwQD65LzP1UZAFYgzQaokBkBMquU/siJswQ==","dtype":"float32","order":"little","shape":[10]},"PC6":{"__ndarray__":"Kb3dQHLSIj9MUVHAEMedvylA8b6yo7G/neXnP6b1D0HcdHLBzAlJQA==","dtype":"float32","order":"little","shape":[10]},"PC7":{"__ndarray__":"HteRwLrMRkEtGXjAMDqQwC67CEGURI++csbPvyq90MDiBHPAqTOGQA==","dtype":"float32","order":"little","shape":[10]},"PC8":{"__ndarray__":"OW58wESzjr+kv3NBwzcgv7ZOekAgnIE8SOo/vzKDsL8MieDAJ5aKwA==","dtype":"float32","order":"little","shape":[10]},"PC9":{"__ndarray__":"XjjeQJ/yHcEVI0q/OmfOwC5rGUGoNT++oecVQHjgq8BRMF8/zgQ7QA==","dtype":"float32","order":"little","shape":[10]},"color":["#8dd3c7","#8dd3c7","#8dd3c7","#8dd3c7","#ffed6f","#ffed6f","#ffed6f","#d6ca66","#d6ca66","#d6ca66"],"coverage":{"__ndarray__":"AAAAAAAA+H8AAAAAAAD4f5nwS/286SBAK01KQbc3JEDogvqWOb0lQFMFo5I6kT9AcEIhAg5BJkAzbf/KShMeQFXZd0XwHyVAkwA1tWxNIkA=","dtype":"float64","order":"little","shape":[10]},"index":[0,1,2,3,4,5,6,7,8,9],"population":["CEU","CEU","CEU","CEU","CHB","CHB","CHB","YRI","YRI","YRI"],"sample":["CEU-0","CEU-1","CEU-2","CEU-3","CHB-0","CHB-1","CHB-2","YRI-0","YRI-1","YRI-2"]},"selected":{"id":"1109"},"selection_policy":{"id":"1108"}},"id":"1059","type":"ColumnDataSource"},{"attributes":{},"id":"1109","type":"Selection"},{"attributes":{"coordinates":null,"group":null,"items":[{"id":"1056"},{"id":"1057"},{"id":"1058"}]},"id":"1055","type":"Legend"},{"attributes":{"tools":[{"id":"1023"},{"id":"1024"},{"id":"1025"},{"id":"1026"},{"id":"1027"},{"id":"1028"},{"id":"1030"}]},"id":"1031","type":"Toolbar"},{"attributes":{"fill_alpha":{"value":0.1},"fill_color":{"field":"color"},"hatch_alpha":{"value":0.1},"hatch_color":{"field":"color"},"line_alpha":{"value":0.1},"size":{"value":15},"x":{"field":"PC1"},"y":{"field":"PC3"}},"id":"1097","type":"Circle"},{"attributes":{"fill_alpha":{"value":0.2},"fill_color":{"field":"color"},"hatch_alpha":{"value":0.2},"hatch_color":{"field":"color"},"line_alpha":{"value":0.2},"size":{"value":15},"x":{"field":"PC1"},"y":{"field":"PC2"}},"id":"1042","type":"Circle"},{"attributes":{"index":4,"label":{"value":"CHB"},"renderers":[{"id":"1099"}]},"id":"1113","type":"LegendItem"},{"attributes":{"tools":[{"id":"1079"},{"id":"1080"},{"id":"1081"},{"id":"1082"},{"id":"1083"},{"id":"1084"},{"id":"1086"}]},"id":"1087","type":"Toolbar"},{"attributes":{"data":{"PC1":{"__ndarray__":"o9RFQdd2BkHQxztBDM2gwQPUWcGPDa3AMfIXwYz0Az/Yet9AAtcJQQ==","dtype":"float32","order":"little","shape":[10]},"PC10":{"__ndarray__":"AAD0tQAAB7YAABq2AAAWtgAAMLYAAMS1AAAUtgAA+LUAwAW2AIACtg==","dtype":"float32","order":"little","shape":[10]},"PC2":{"__ndarray__":"o4nNwJpKhsCgvqu+azLQwHP1BcE2vL5B+uo7QLY7dcB+rpjAehP1QA==","dtype":"float32","order":"little","shape":[10]},"PC3":{"__ndarray__":"0d3BwIKKT79jjDJAoLw1vhO43cCH9SrBlD+WQf6eqsD/jJG/9u4YQQ==","dtype":"float32","order":"little","shape":[10]},"PC4":{"__ndarray__":"XTNWwaPDi8BH7VC/oU4bwRgexEDJaR7ABF5Lvy7agUEc2KJABieDQA==","dtype":"float32","order":"little","shape":[10]},"PC5":{"__ndarray__":"O8siQHqPpUAwU/u//vYSwQD65LzP1UZAFYgzQaokBkBMquU/siJswQ==","dtype":"float32","order":"little","shape":[10]},"PC6":{"__ndarray__":"Kb3dQHLSIj9MUVHAEMedvylA8b6yo7G/neXnP6b1D0HcdHLBzAlJQA==","dtype":"float32","order":"little","shape":[10]},"PC7":{"__ndarray__":"HteRwLrMRkEtGXjAMDqQwC67CEGURI++csbPvyq90MDiBHPAqTOGQA==","dtype":"float32","order":"little","shape":[10]},"PC8":{"__ndarray__":"OW58wESzjr+kv3NBwzcgv7ZOekAgnIE8SOo/vzKDsL8MieDAJ5aKwA==","dtype":"float32","order":"little","shape":[10]},"PC9":{"__ndarray__":"XjjeQJ/yHcEVI0q/OmfOwC5rGUGoNT++oecVQHjgq8BRMF8/zgQ7QA==","dtype":"float32","order":"little","shape":[10]},"color":["#8dd3c7","#8dd3c7","#8dd3c7","#8dd3c7","#ffed6f","#ffed6f","#ffed6f","#d6ca66","#d6ca66","#d6ca66"],"coverage":{"__ndarray__":"AAAAAAAA+H8AAAAAAAD4f5nwS/286SBAK01KQbc3JEDogvqWOb0lQFMFo5I6kT9AcEIhAg5BJkAzbf/KShMeQFXZd0XwHyVAkwA1tWxNIkA=","dtype":"float64","order":"little","shape":[10]},"index":[0,1,2,3,4,5,6,7,8,9],"population":["CEU","CEU","CEU","CEU","CHB","CHB","CHB","YRI","YRI","YRI"],"sample":["CEU-0","CEU-1","CEU-2","CEU-3","CHB-0","CHB-1","CHB-2","YRI-0","YRI-1","YRI-2"]},"selected":{"id":"1165"},"selection_policy":{"id":"1164"}},"id":"1115","type":"ColumnDataSource"},{"attributes":{},"id":"1048","type":"AllLabels"},{"attributes":{"source":{"id":"1059"}},"id":"1100","type":"CDSView"},{"attributes":{"coordinates":null,"data_source":{"id":"1059"},"glyph":{"id":"1096"},"group":null,"hover_glyph":null,"muted_glyph":{"id":"1098"},"nonselection_glyph":{"id":"1097"},"view":{"id":"1100"}},"id":"1099","type":"GlyphRenderer"},{"attributes":{"fill_alpha":{"value":0.2},"fill_color":{"field":"color"},"hatch_alpha":{"value":0.2},"hatch_color":{"field":"color"},"line_alpha":{"value":0.2},"size":{"value":15},"x":{"field":"PC1"},"y":{"field":"PC3"}},"id":"1098","type":"Circle"},{"attributes":{},"id":"1047","type":"BasicTickFormatter"},{"attributes":{},"id":"1051","type":"AllLabels"},{"attributes":{"coordinates":null,"group":null,"items":[{"id":"1112"},{"id":"1113"},{"id":"1114"}]},"id":"1111","type":"Legend"},{"attributes":{},"id":"1053","type":"Selection"},{"attributes":{},"id":"1107","type":"AllLabels"},{"attributes":{"source":{"id":"1003"}},"id":"1044","type":"CDSView"},{"attributes":{"coordinates":null,"data_source":{"id":"1003"},"glyph":{"id":"1040"},"group":null,"hover_glyph":null,"muted_glyph":{"id":"1042"},"nonselection_glyph":{"id":"1041"},"view":{"id":"1044"}},"id":"1043","type":"GlyphRenderer"},{"attributes":{},"id":"1139","type":"ResetTool"},{"attributes":{},"id":"1050","type":"BasicTickFormatter"},{"attributes":{"fill_alpha":{"value":0.1},"fill_color":{"field":"color"},"hatch_alpha":{"value":0.1},"hatch_color":{"field":"color"},"line_alpha":{"value":0.1},"size":{"value":15},"x":{"field":"PC1"},"y":{"field":"PC2"}},"id":"1041","type":"Circle"},{"attributes":{"index":0,"label":{"value":"CEU"},"renderers":[{"id":"1099"}]},"id":"1112","type":"LegendItem"},{"attributes":{},"id":"1135","type":"PanTool"},{"attributes":{},"id":"1106","type":"BasicTickFormatter"},{"attributes":{},"id":"1125","type":"LinearScale"},{"attributes":{},"id":"1119","type":"DataRange1d"},{"attributes":{},"id":"1052","type":"UnionRenderers"},{"attributes":{"below":[{"id":"1127"}],"center":[{"id":"1130"},{"id":"1134"},{"id":"1167"}],"height":400,"left":[{"id":"1131"}],"renderers":[{"id":"1155"}],"right":[{"id":"1167"}],"title":{"id":"1117"},"toolbar":{"id":"1143"},"toolbar_location":null,"width":400,"x_range":{"id":"1119"},"x_scale":{"id":"1123"},"y_range":{"id":"1121"},"y_scale":{"id":"1125"}},"id":"1116","subtype":"Figure","type":"Plot"},{"attributes":{},"id":"1138","type":"SaveTool"},{"attributes":{"index":7,"label":{"value":"YRI"},"renderers":[{"id":"1099"}]},"id":"1114","type":"LegendItem"},{"attributes":{"overlay":{"id":"1141"}},"id":"1137","type":"BoxZoomTool"},{"attributes":{"bottom_units":"screen","coordinates":null,"fill_alpha":0.5,"fill_color":"lightgrey","group":null,"left_units":"screen","level":"overlay","line_alpha":1.0,"line_color":"black","line_dash":[4,4],"line_width":2,"right_units":"screen","syncable":false,"top_units":"screen"},"id":"1029","type":"BoxAnnotation"},{"attributes":{"coordinates":null,"group":null,"text":"PC2 vs PC3"},"id":"1117","type":"Title"},{"attributes":{},"id":"1140","type":"HelpTool"},{"attributes":{},"id":"1104","type":"AllLabels"},{"attributes":{"axis":{"id":"1131"},"coordinates":null,"dimension":1,"group":null,"ticker":null},"id":"1134","type":"Grid"},{"attributes":{"callback":null,"tooltips":[["sample","@sample"],["population","@population"],["coverage","@coverage"]]},"id":"1142","type":"HoverTool"},{"attributes":{},"id":"1132","type":"BasicTicker"},{"attributes":{"fill_alpha":{"value":0.8},"fill_color":{"field":"color"},"hatch_alpha":{"value":0.8},"hatch_color":{"field":"color"},"line_alpha":{"value":0.8},"size":{"value":15},"x":{"field":"PC2"},"y":{"field":"PC3"}},"id":"1152","type":"Circle"},{"attributes":{"axis_label":"PC3 (13.0%)","coordinates":null,"formatter":{"id":"1159"},"group":null,"major_label_policy":{"id":"1160"},"ticker":{"id":"1132"}},"id":"1131","type":"LinearAxis"},{"attributes":{},"id":"1103","type":"BasicTickFormatter"},{"attributes":{},"id":"1136","type":"WheelZoomTool"},{"attributes":{},"id":"1123","type":"LinearScale"},{"attributes":{"axis":{"id":"1127"},"coordinates":null,"group":null,"ticker":null},"id":"1130","type":"Grid"},{"attributes":{"axis_label":"PC2 (16.1%)","coordinates":null,"formatter":{"id":"1162"},"group":null,"major_label_policy":{"id":"1163"},"ticker":{"id":"1128"}},"id":"1127","type":"LinearAxis"},{"attributes":{},"id":"1121","type":"DataRange1d"},{"attributes":{},"id":"1128","type":"BasicTicker"}],"root_ids":["1207"]},"title":"Bokeh Application","version":"2.4.3"}};
  const render_items = [{"docid":"f4e390b0-9777-4446-8ec9-182b8a0276c5","root_ids":["1207"],"roots":{"1207":"b1fbb2f8-0bdc-4c89-b2f2-c618e9aa6f56"}}];
  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);
  }
  if (root.Bokeh !== undefined) {
    embed_document(root);
  } else {
    let attempts = 0;
    const timer = setInterval(function(root) {
      if (root.Bokeh !== undefined) {
        clearInterval(timer);
        embed_document(root);
      } else {
        attempts++;
        if (attempts > 100) {
          clearInterval(timer);
          console.log("Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing");
        }
      }
    }, 10, root)
  }
})(window);</script></div>
</div>
</section>
<div class="toctree-wrapper compound">
</div>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "pgip"
        },
        kernelOptions: {
            kernelName: "pgip",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'pgip'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="describing_variation.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Describing variation</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="pca/sgkit.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">PCA with sgkit</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Per Unneberg<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>